proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -371,18 +374,393 @@ lemma decode_sc_ctrl_inv_corres:
+                          apply (frule (1) sc_ko_at_valid_objs_valid_sc'[rotated])
+                          apply (clarsimp simp: valid_sched_context'_def valid_sched_context_size'_def
+                                                obj_at'_def projectKOs)
+                          apply (erule_tac x=tp in valid_objsE', simp)
+                          apply (fastforce simp: valid_obj'_def valid_tcb'_def obj_at'_def projectKOs
+                                                 inQ_def tcb_cte_cases_def objBits_simps' scBits_simps)
+                         apply (clarsimp simp: tcb_relation_def)
+                        apply (wpsimp wp: hoare_vcg_if_lift2 hoare_drop_imp)
+                       apply (wpsimp wp: hoare_vcg_if_lift2 hoare_drop_imp)
+                      apply wpsimp
+                     apply wpsimp
+                    apply (rule_tac P="?abs_buf and sc_yf_sc_at ((=) None) scp and ?scp
+                                       and (\<lambda>s. sched = is_schedulable_bool tp s)"
+                                and P'="?con_buf and cur_tcb' and ko_at' sc0' scp"
+                           in corres_inst)
+                    apply simp
+                   apply (intro conjI impI; clarsimp cong: conj_cong)
+                   apply (clarsimp simp: invs_def valid_state_def valid_pspace_def valid_sched_def
+                                         valid_sched_action_weak_valid_sched_action pred_tcb_at_def
+                                         sc_tcb_sc_at_def obj_at_def is_tcb is_sc_obj)
+                   apply (fastforce simp: obj_at_def elim!: valid_sched_context_size_objsI)
+                  apply (clarsimp cong: conj_cong simp: invs'_def valid_pspace'_def cur_tcb'_def)
+                 apply (clarsimp simp:obj_at'_def projectKOs)
+                 apply (wpsimp wp: thread_get_wp hoare_case_option_wp)+
+                   apply (wpsimp wp: threadGet_wp)+
+               apply (wpsimp wp: is_schedulable_wp)
+              apply (wpsimp wp: isSchedulable_wp)
+             apply (clarsimp simp: invs_def valid_state_def valid_pspace_def valid_objs_valid_tcbs
+                             cong: conj_cong imp_cong if_cong)
+             apply (rule context_conjI;
+                    clarsimp simp: obj_at_def sc_tcb_sc_at_def is_sc_obj
+                            elim!: valid_sched_context_size_objsI dest!: invs_valid_objs)
+              apply (fastforce simp: valid_obj_def obj_at_def valid_sched_context_def)
+             apply (intro conjI impI; clarsimp simp: is_tcb pred_tcb_at_def obj_at_def)
+              apply (erule (1) valid_sched_context_size_objsI)+
+            apply (clarsimp simp: invs'_def valid_pspace'_def valid_objs'_valid_tcbs')
+            apply (clarsimp simp: obj_at'_def projectKOs opt_map_red)
+           apply (rule_tac Q="\<lambda>rv. (case buf of None \<Rightarrow> \<lambda>_. True | Some x \<Rightarrow> in_user_frame x) and einvs and
```

Comment:
It doesn't make much of a difference, but I guess that these `hoare_strengthen_posts` could use `case_option` in the same way the lemma preconditions do.

Reply:
These big strengthening patterns are typically copied over from the output pane, and this is how Isabelle prints it.